dot/sync/chain_processor_test.go

```diff
@@ -0,0 +1,607 @@
+		Body: types.Body{},
+	}
+
+	mockBlockState := NewMockBlockState(ctrl)
+	mockHeader := &types.Header{
+		Number:    0,
+		StateRoot: trie.EmptyHash,
+	}
+	mockHeaderHash := mockHeader.Hash()
+	mockBlock := &types.Block{
+		Header: types.Header{
+			Number: 0,
+		},
+		Body: types.Body{},
+	}
+	mockTrieState, err := storage.NewTrieState(nil)
+	require.NoError(t, err)
+	mockBlockState.EXPECT().GetHeader(common.Hash{}).Return(mockHeader, nil)
+	mockInstance := mocks.NewMockInstance(ctrl)
+	mockInstance.EXPECT().SetContextStorage(mockTrieState)
+	mockInstance.EXPECT().ExecuteBlock(mockBlock)
+	mockBlockState.EXPECT().GetRuntime(&mockHeaderHash).Return(mockInstance, nil)
+
+	mockStorageState := NewMockStorageState(ctrl)
+	mockStorageState.EXPECT().Lock()
+	mockStorageState.EXPECT().Unlock()
+	mockStorageState.EXPECT().TrieState(&trie.EmptyHash).Return(mockTrieState, nil)
+
+	mockBlockImportHandler := NewMockBlockImportHandler(ctrl)
+	mockBlockImportHandler.EXPECT().HandleBlockImport(mockBlock, mockTrieState).Return(nil)
+
+	mockTelemetry := NewMockClient(ctrl)
+	mockTelemetry.EXPECT().SendMessage(gomock.Any())
+
+	s := &chainProcessor{
+		blockState:         mockBlockState,
+		storageState:       mockStorageState,
+		blockImportHandler: mockBlockImportHandler,
+		telemetry:          mockTelemetry,
+	}
+	err = s.handleBlock(block)
+	assert.NoError(t, err)
+}
+
+func Test_chainProcessor_handleBlock_nilBlock(t *testing.T) {
+	t.Parallel()
+
+	s := &chainProcessor{}
+	err := s.handleBlock(nil)
+	assert.EqualError(t, err, errors.New("block or body is nil").Error())
+}
+
+func Test_chainProcessor_handleBody(t *testing.T) {
+	t.Parallel()
+
+	var testExtrinsic = []types.Extrinsic{{1, 2, 3}, {7, 8, 9, 0}, {0xa, 0xb}}
+
+	ctrl := gomock.NewController(t)
+
+	mockTransactionState := NewMockTransactionState(ctrl)
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[0])
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[1])
+	mockTransactionState.EXPECT().RemoveExtrinsic(testExtrinsic[2])
+
+	testBody := types.NewBody(testExtrinsic)
+
+	type fields struct {
+		transactionState TransactionState
+	}
+	type args struct {
+		body *types.Body
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+	}{
+		{
+			name: "base case",
+			fields: fields{
+				transactionState: mockTransactionState,
+			},
+			args: args{body: testBody},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := &chainProcessor{
+				transactionState: tt.fields.transactionState,
+			}
+			s.handleBody(tt.args.body)
+		})
+	}
+}
+
+func Test_chainProcessor_handleHeader(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+
+	mockBabeVerifier := NewMockBabeVerifier(ctrl)
+	mockBabeVerifier.EXPECT().VerifyBlock(gomock.AssignableToTypeOf(&types.Header{})).DoAndReturn(func(h *types.
+		Header) error {
+		if h == nil {
+			return errors.New("nil header")
+		}
+		return nil
+	}).Times(2)
```

Comment:
Configure this in the subtest either with a test case mock builder function, or with some exepcted argument and return value.

Also avoid `gomock.AssignableToTypeOf(&types.Header{})` and `DoAndReturn` and expect concrete arguments/return pre-defined values without additional logic.

You'll also have to remove `Times(2)` I guess since you won't re-use the mock anymore.

Reply:
Updated to use mock builder function.